Client/Frontend/UserContent/UserScripts/Scripts_Dynamic/ScriptHandlers/Paged/SolanaProviderScriptHandler.swift

```diff
@@ -0,0 +1,344 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import WebKit
+import BraveCore
+import BraveShared
+import os.log
+
+class SolanaProviderScriptHandler: TabContentScript {
+  
+  fileprivate enum Keys: String {
+    case connect
+    case disconnect
+    case signMessage
+    case message
+    case serializedMessage
+    case signatures
+    case sendOptions
+    case publicKey
+    case signature
+    case method
+    case params
+    case code
+    case data
+  }
+  
+  static let scriptName = "WalletSolanaProviderScript"
+  static let scriptId = UUID().uuidString
+  static let messageHandlerName = "\(scriptName)_\(messageUUID)"
+  static let scriptSandbox: WKContentWorld = .page
+  static let userScript: WKUserScript? = {
+    guard var script = loadUserScript(named: scriptName) else {
+      return nil
+    }
+    return WKUserScript.create(source: secureScript(handlerName: messageHandlerName,
+                                                    securityToken: scriptId,
+                                                    script: script),
+                               injectionTime: .atDocumentStart,
+                               forMainFrameOnly: true,
+                               in: scriptSandbox)
+  }()
+  
+  private weak var tab: Tab?
+  
+  init(tab: Tab) {
+    self.tab = tab
+  }
+  
+  static func shouldInjectWalletProvider(_ completion: @escaping (Bool) -> Void) {
+    BraveWallet.KeyringServiceFactory.get(privateMode: false)?
+      .keyringInfo(BraveWallet.SolanaKeyringId, completion: { keyring in
+        completion(keyring.isKeyringCreated)
+      })
+  }
+  
+  private struct MessageBody: Decodable {
+    enum Method: String, Decodable {
+      case connect
+      case disconnect
+      case signAndSendTransaction
+      case signMessage
+      case request
+      case signTransaction
+      case signAllTransactions
+    }
+    var method: Method
+    var args: String?
+    
+    private enum CodingKeys: String, CodingKey {
+      case method
+      case args
+    }
+  }
+  
+  func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage, replyHandler: @escaping (Any?, String?) -> Void) {
+    Task { @MainActor in
```

Comment:
Task is created for the entire body but the first await doesnt happen until you execute on the body method. To avoid more `Sendable` issues down the road, the Task should only surround the async required parts

Reply:
Fixed in [542fba8](https://github.com/brave/brave-ios/pull/6050/commits/542fba84d52d2bee4972af100d529aeef7d6532c)